### PR TITLE
Added Query feature

### PR DESCRIPTION
**Pull request by Humbarrt**
Hi there,

I wanted the clock to be able to utilize the query as input. Since other people were asking for the same feature (#168), I figured I might take a shot at it. My implementation isn't pretty, but it's a start. This is my first time working on a Grafana plugin. 
Have a great day!

### CODE REVIEW

File: src/options.ts
Comment (academo) [inline src/options.ts:135]: You can use instead `addFieldNamePicker` from `grafana-data` (and in the count up)

File: src/types.ts
Comment (academo) [inline src/types.ts:20]: these two types are almost the same. extract the common part. e.g; `QueryCalculation`

File: src/ClockPanel.tsx
Comment (academo) [inline src/ClockPanel.tsx:69]: you are passing all the data and responsibility of calculate your field to a component that should merely render data. Instead, create another function that abstracts all the logic and returns the field you actually want and then pass it here.

Comment (academo) [reply src/ClockPanel.tsx:69]: you also want a loading state. since you can only use data when it is done. You will want to show a [spinner](https://developers.grafana.com/ui/latest/index.html?path=/story/visualizations-spinner--basic) while the data loads when the source is a query or it'll show blank while it loads.

File: src/components/RenderTime.tsx
Comment (academo) [inline src/components/RenderTime.tsx:44]: what would it display when there are not data series and the state is done? 
There should be feedback for the user indicating that the clock can't display because there's not data.

File: src/components/RenderTime.tsx
Comment (academo) [inline src/components/RenderTime.tsx:46]: nit: you can reduce all this condition to `!field_name?.length || !data?.series?.[0]?.fields?.length`

File: src/components/RenderTime.tsx
Comment (academo) [inline src/components/RenderTime.tsx:68]: all this logic should be extracted to another function and file.

File: src/components/RenderTime.tsx
Comment (academo) [inline src/components/RenderTime.tsx:70]: please do not use the `any` type

File: src/components/RenderTime.tsx
Comment (academo) [inline src/components/RenderTime.tsx:175]: all this logic is duplicated.  only the return text changes.

**Review by academo — CHANGES_REQUESTED** [review]
Thanks so much for your contribution!

You've done great for your first time working in a grafana plugin.

There are several things I'd like you to work on before we merge this PR. Most of them go around code orgnanization and de-duplication rather than the functionality which looks good to me.

There's an important part on user feedback when there's no suitable data to show. For example, right now it'll just show the end text in situations where there's no suitable data to render, instead it should tell the user that the data can not be used for a clock. Think about the user experience and what would you like, as an user, to see if there's something wrong in the data you are using.

Give it a try, throw data at it that makes no sense, something an user might just do, and see what happens. Try several data series too.

**Comment by Humbarrt** [comment]
Hi,
I changed things according to your requests. I added a No Value Text and a Invalid Value Text, so if a user wants to debug their query, they can. In production you might want an empty field, thus I made it changeable. I also added a description under the countdown/countup so that you can have a label for the query result... something like:
`{
'name': ['Next New Year', 'Last New Year', 'Year After Next Year'],
'event': ['2025-01-01', '2024-01-01', '2026-01-01']
}`
where the 'name' could be the description and the 'event' could be the countup/countdown time. The description can also be static or not be displayed at all.

In my tests i found that all normal time inputs are working, except the epoch input. But a user could fix that with a grafana data transform to multiply the row by 1000, so I think it's no big deal. In the future you maybe want to consider a custom input format though. 

Anyway, I know ComputeTime.tsx isn't super pretty, since it handles both time and description, maybe you have a better idea on how to handle that. 
Cheers!

File: src/ClockPanel.tsx
Comment (Humbarrt) [reply src/ClockPanel.tsx]: I dont know if you want to show a spinner every time the data loads. Grafana itself shows if a query is done and this logic isn't taking that much more time.

File: src/components/RenderTime.tsx
Comment (academo) [inline src/components/RenderTime.tsx]: even though this makes sense it is best to not change this existing behaviour or people will get unexpected behaviours when they update their plugin.

File: src/ClockPanel.tsx
Comment (academo) [inline src/ClockPanel.tsx:58]: I was rather confused with this variable and the `getTime` function. I now understand this `time` is the "targetTime". The time the counter will go up or down. Perhaps you should call it as such.

Also it is ok if you want to get the `description` from this same function but you can do two things:

* move this function out of the ComputeTime.tsx component. Since that is a react component and there's no reason to have it there. move it to its own function and perhaps you can rename it to something such as `calculateClockOptions` or `getProcessedClockOptions`. something that tells this function is doing calculations based on the options it gets.

File: src/options.ts
Comment (academo) [inline src/options.ts:323]: To avoid this getting confusing let's not call this "description" you can maybe it called `descriptionText` or `descriptionContent` otherwise it is not clear what is inside this in the rest of the code.

File: .config/.eslintrc
Comment (academo) [inline .config/.eslintrc:18]: Is there any particular reason you are changing this to a warning?

**Review by academo — CHANGES_REQUESTED** [review]
Hi @Humbarrt I really like the recent changes to consolidate all the logic to calculate the target time and the description addition. though I would had rather you put the description part in a separate PR to make this review faster, since it is not core to the query functionality. You can keep it here for now since I already reviewed. 

I added some comments, many of them are nits about variable names and code style changes to make the code more legible but the core logic seems good to go once you solve the problem with multiple series. I have also tested the plugin locally and it works just as expected with the backwards compatibility we need with previous versions. 

before we merge it, add unit tests for the cases you created.

 You updated the tests so the types work but you didn't add test cases for displaying the data based on a series of data.  This is important so we know the plugin is working in future updates. I recommend you the following test cases to add:

* Count up from query field
* Count down from query field
* Description with static text
* Description from query field
* Finished count up from query field
* Finished count down from query field
* Error state for count up/down
* Error state for description

File: src/ClockPanel.tsx
Comment (academo) [reply src/ClockPanel.tsx]: you are right! Let's just leave it as is :)

File: .config/.eslintrc
Comment (Humbarrt) [reply .config/.eslintrc]: yes, when i changed the types as you requested
> these two types are almost the same. extract the common part. e.g; `QueryCalculation`

I used const/type instead of enum, but I got an error for defining it twice. That's why I changed it to be a warning:
clock-panel/src/types.ts
40:13  warning  'CountdownQueryCalculation' is already defined  no-redeclare
46:13  warning  'CountupQueryCalculation' is already defined    no-redeclare

I agree that I shouldn't have disabled it overall. Now I disable it by line and I've removed the line from [.config/.eslintrc](https://github.com/grafana/clock-panel/pull/179/files/39e5c0d73a36297329823703140727089ec3fd03#diff-71f3678a7bd77030e1c7ecacee33c4b70ba99e6e7f0c1dfe0243ef3e700d6e07)

File: src/components/ComputeTime.tsx
Comment (Humbarrt) [reply src/components/ComputeTime.tsx:130]: I only did that because you do the same in [RenderZone.tsx](https://github.com/grafana/clock-panel/blame/e342be1aa37ea63d42aa6631448b863169550a2b/src/components/RenderZone.tsx#L37) - also as a fallback in a switch statement. My console.log should not happen, it's only triggering if someone adds QueryCalculations without adding cases to the switch.

Comment (Humbarrt) [reply src/components/ComputeTime.tsx:130]: I only did that because you did the same in [RenderZone.tsx](https://github.com/grafana/clock-panel/blame/e342be1aa37ea63d42aa6631448b863169550a2b/src/components/RenderZone.tsx#L37) - also as a fallback in a switch statement. My console.log should not happen, it's only triggered if someone adds QueryCalculations without adding cases to the switch.

File: src/components/ComputeTime.tsx
Comment (Humbarrt) [reply src/components/ComputeTime.tsx]: We use the fieldValues for the QueryCalculation.last &  QueryCalculation.first finalValue and it should not be filtered for that. But I can put 
`let values: QueryRow[] = fieldValues
        .filter((v) => v.time !== null && v.time !== undefined && !Number.isNaN(v.time))
        .map((v) => {
          return { time: moment(v.time), description: v.description };
        });`
into a single loop. Good nit :)

**Comment by Humbarrt** [comment]
Hi, I think I addressed all your requested changes. Please let me know if you find other problems.
I will do the tests next.
Cheers!

File: src/components/CalculateClockOptions.ts
Comment (academo) [inline src/components/CalculateClockOptions.ts:67]: it is possible that we have several series data but some of them came empty so having no fields in series[0] doesn't mean the other don't have data.

**Comment by academo** [comment]
> Hi, I think I addressed all your requested changes. Please let me know if you find new problems. I will do the tests next. Cheers!

This looks almost ready to do. looking forward for the test and merging this :D

File: src/components/CalculateClockOptions.ts
Comment (Humbarrt) [reply src/components/CalculateClockOptions.ts:67]: Ok, I will remove the `isDataSeriesPopulated` condition as empty fields would be catched by the `timeField` not finding anything.

**Comment by academo** [comment]
> Hi, while creating the tests I stumbled upon something. The showDate setting renders the current date and not the targetTime. When I do a countdown/countup it would be useful to know what date I am counting down/up to. This would be more intuitive, at least for me. Now, I know this is only how I feel and changing it would alter existing behavior for users. I just wanted to bring it up. Cheers!

It is best not to change any of the existing behaviour even if that behaviour is somewhat counter intuitive because there are many plugin users that rely on it. Clock panel is rather a popular plugin in Grafana and we don't want to break the workflow of the existing users.

**Comment by Humbarrt** [comment]
Ok, I thought so. As I said, I just wanted to bring it up. Maybe we can add an option to the ShowDate later. 

I've done the tests you asked for. Maybe I'll add some more tomorrow. E.g:
Time/Description from different Series
Null as input
Undefined as input
NaN as input
different modes: lastNotNull, last, firstNotNull, first, min, minFuture, max and maxPast

Cheers!

**Comment by Humbarrt** [comment]
Hi, I've done all tests I can think of. If there is anything else I can do, let me know :)
Cheers!

File: src/plugin.json
Comment (tolzhabayev) [inline src/plugin.json:20]: Please avoid formatting changes that are not relevant to the code change. In this case it would be everything besides `skipDataQuery`

File: src/ClockPanel.tsx
Comment (tolzhabayev) [inline src/ClockPanel.tsx:11]: What is the purpose of changing this import?

On the line 59 you can just use `Moment`

**Comment by tolzhabayev** [comment]
The description settings are confusing to the users as they are only relevant for `countdown` and `countup` but are still displayed for any mode.

**Comment by tolzhabayev** [comment]
Generally we should add this case to the testing dashboard https://github.com/grafana/clock-panel/tree/main/provisioning/dashboards - maybe together with a testdata datasource to showcase different behaviour.

File: src/ClockPanel.tsx
Comment (Humbarrt) [reply src/ClockPanel.tsx]: Yes, you are right!

**Comment by Humbarrt** [comment]
> The description settings are confusing to the users as they are only relevant for `countdown` and `countup` but are still displayed for any mode. The setting should only be visible when relevant.

Imo the description should be available in all modes. But you were right, there was a bug that showed `query` as a description source for the time mode. I fixed that. Now you can have none/input description source for time-mode, countup/countdown without query and none/input/query if you use a query for you countup/countdown

**Comment by Humbarrt** [comment]
> Generally we should add this case to the testing dashboard https://github.com/grafana/clock-panel/tree/main/provisioning/dashboards - maybe together with a testdata datasource to showcase different behaviour.

Hi, I added panels to the testing dashboard and fixed a bug that I found in when using the older grafana version.
Cheers!

**Review by tolzhabayev — APPROVED** [review]
@Humbarrt thanks for making the changes and for contributing!

**Review by academo — APPROVED** [review]
@Humbarrt thanks for such a great contribution!

I will merge and release this next week.

**Pull request merged**
